kvstreamer: nil transaction causes panic in kvstreamer.NewStreamer

roachtest.sqlsmith/setup=tpch-sf1/setting=default [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_RoachtestNightlyGceBazel/5718837?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_RoachtestNightlyGceBazel/5718837?buildTab=artifacts#/sqlsmith/setup=tpch-sf1/setting=default) on master @ [571bfa3afb3858ae84d8a8fcdbb0a38e058402a5](https://github.com/cockroachdb/cockroach/commits/571bfa3afb3858ae84d8a8fcdbb0a38e058402a5):


```
								(NULL),
								(
									ARRAY[832826517:::OID,336161718:::OID,3712606969:::OID,1868633338:::OID,2863025740:::OID,704314154:::OID]
								),
								((SELECT ARRAY[2437433960:::OID,740186049:::OID,3595763864:::OID] AS col_48 LIMIT 1:::INT8))
						)
							AS tab_28 (col_49)
				),
			with_8 (col_50)
				AS (
					SELECT
						*
					FROM
						(VALUES ('11:36:36.683263+02:11:00':::TIMETZ), ('07:46:55.783482-00:37:00':::TIMETZ), (NULL))
							AS tab_29 (col_50)
				)
		SELECT
			NULL AS col_51,
			(tab_31.n_nationkey::INT8 # NULL::INT8)::INT8 AS col_52,
			tab_31.crdb_internal_mvcc_timestamp AS col_53,
			e'\x0eF\x0fX<\x0b':::STRING AS col_54,
			tab_31.n_nationkey AS col_55
		FROM
			[
				INSERT
				INTO
					defaultdb.public.customer AS tab_30
						(c_address, c_name, c_comment, c_mktsegment, c_acctbal, c_phone, c_nationkey, c_custkey)
				VALUES
					(
						e'\x15':::STRING,
						e'ukbDI\n8':::STRING,
						'52465':::STRING,
						'``':::STRING,
						0.3671557088061544:::FLOAT8,
						'[':::STRING,
						(-2047104154723331035):::INT8,
						7862657228055256243:::INT8
					)
				RETURNING
					']':::STRING AS ret_1
			],
			defaultdb.public.nation@[0] AS tab_31
		WHERE
			true
		ORDER BY
			tab_31.n_nationkey, tab_31.n_nationkey ASC, tab_31.n_comment DESC
		LIMIT
			87:::INT8;ping node 1: dial tcp 34.74.155.56:26257: connect: connection refused
		HINT: node likely crashed, check logs in artifacts > logs/1.unredacted
```
<p>Parameters: <code>ROACHTEST_cloud=gce</code>
, <code>ROACHTEST_cpu=4</code>
, <code>ROACHTEST_ssd=0</code>
</p>
<details><summary>Help</summary>
<p>


See: [roachtest README](https://github.com/cockroachdb/cockroach/blob/master/pkg/cmd/roachtest/README.md)



See: [How To Investigate \(internal\)](https://cockroachlabs.atlassian.net/l/c/SSSBr8c7)

</p>
</details>
/cc @cockroachdb/sql-queries
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*sqlsmith/setup=tpch-sf1/setting=default.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>


Jira issue: CRDB-17539

Looks like a panic in the streamer:
```
panic: runtime error: invalid memory address or nil pointer dereference [recovered]
	panic: runtime error: invalid memory address or nil pointer dereference [recovered]
	panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x8 pc=0x21f7ce6]

goroutine 2548 [running]:
panic({0x4b6aa20, 0xc0044da780})
	GOROOT/src/runtime/panic.go:1147 +0x3a8 fp=0xc002540930 sp=0xc002540870 pc=0x48a828
github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).closeWrapper(0xc004ca0000, {0x6755af0, 0xc004d46f00}, {0x48e0be0, 0x95d29c0})
	github.com/cockroachdb/cockroach/pkg/sql/conn_executor.go:1076 +0x1b4 fp=0xc0025409c0 sp=0xc002540930 pc=0x3222ab4
github.com/cockroachdb/cockroach/pkg/sql.(*Server).ServeConn.func1({0x6755af0, 0xc004d46f00}, {0xac3f909ae})
	github.com/cockroachdb/cockroach/pkg/sql/conn_executor.go:776 +0x4a fp=0xc002540a00 sp=0xc0025409c0 pc=0x344512a
github.com/cockroachdb/cockroach/pkg/sql.(*Server).ServeConn·dwrap·27()
	github.com/cockroachdb/cockroach/pkg/sql/conn_executor.go:777 +0x35 fp=0xc002540a28 sp=0xc002540a00 pc=0x3220d55
runtime.deferCallSave(0xc002540b30, 0xc002547df0)
	GOROOT/src/runtime/panic.go:950 +0x82 fp=0xc002540a38 sp=0xc002540a28 pc=0x48a422
runtime.runOpenDeferFrame(0xc001b00280, 0xc003429ae0)
	GOROOT/src/runtime/panic.go:889 +0x27b fp=0xc002540ab8 sp=0xc002540a38 pc=0x489dbb
panic({0x48e0be0, 0x95d29c0})
	GOROOT/src/runtime/panic.go:1038 +0x215 fp=0xc002540b78 sp=0xc002540ab8 pc=0x48a695
github.com/cockroachdb/cockroach/pkg/sql/colexecerror.CatchVectorizedRuntimeError.func1()
	github.com/cockroachdb/cockroach/pkg/sql/colexecerror/error.go:58 +0x3bb fp=0xc002540d10 sp=0xc002540b78 pc=0x201787b
runtime.deferCallSave(0xc002540e18, 0xc002544eb0)
	GOROOT/src/runtime/panic.go:950 +0x82 fp=0xc002540d20 sp=0xc002540d10 pc=0x48a422
runtime.runOpenDeferFrame(0xc00a86e380, 0xc001b00280)
	GOROOT/src/runtime/panic.go:889 +0x27b fp=0xc002540da0 sp=0xc002540d20 pc=0x489dbb
panic({0x48e0be0, 0x95d29c0})
	GOROOT/src/runtime/panic.go:1038 +0x215 fp=0xc002540e60 sp=0xc002540da0 pc=0x48a695
runtime.panicmem(...)
	GOROOT/src/runtime/panic.go:221
runtime.sigpanic()
	GOROOT/src/runtime/signal_unix.go:735 +0x327 fp=0xc002540eb0 sp=0xc002540e60 pc=0x4a1247
github.com/cockroachdb/cockroach/pkg/kv/kvclient/kvstreamer.NewStreamer(0x460ef4, 0xc002541030, 0x30, 0xc000ac0000, 0x5068a35, 0x100000000000000, 0x7f32eddeed90, 0x28, 0x28)
	github.com/cockroachdb/cockroach/pkg/kv/kvclient/kvstreamer/streamer.go:366 +0x26 fp=0xc002540f90 sp=0xc002540eb0 pc=0x21f7ce6
github.com/cockroachdb/cockroach/pkg/sql/row.NewStreamingKVFetcher(0x0, 0x0, 0x0, 0x0, 0xf0f6ac00, 0x1ee, 0xc0034297c0, 0x0, 0x0, 0x1, ...)
	github.com/cockroachdb/cockroach/pkg/sql/row/kv_fetcher.go:143 +0xc5 fp=0xc002541040 sp=0xc002540f90 pc=0x22134c5
github.com/cockroachdb/cockroach/pkg/sql/rowexec.newJoinReader(0xc0029b4f00, 0x67d4ea8, 0xc003347380, {0x677ec38, 0xc0029b50e0}, 0xc0034297c0, {0x0, 0x0}, 0x0)
	github.com/cockroachdb/cockroach/pkg/sql/rowexec/joinreader.go:488 +0x15d5 fp=0xc002541390 sp=0xc002541040 pc=0x245bc55
github.com/cockroachdb/cockroach/pkg/sql/rowexec.NewProcessor({0x6755af0, 0xc00217b340}, 0x7f33182d5a68, 0x10, 0xc002b032b0, 0xc0053e6630, {0xc0053e6630, 0xc001a3a3c0, 0xc002543548}, {0xc0053e6640, ...}, ...)
	github.com/cockroachdb/cockroach/pkg/sql/rowexec/processors.go:154 +0x2337 fp=0xc002543470 sp=0xc002541390 pc=0x246b097
github.com/cockroachdb/cockroach/pkg/sql/colexec/colbuilder.opResult.createAndWrapRowSource.func1({0xc0053e6630, 0x1, 0x1})
	github.com/cockroachdb/cockroach/pkg/sql/colexec/colbuilder/execplan.go:551 +0x135 fp=0xc002543558 sp=0xc002543470 pc=0x3143e15
github.com/cockroachdb/cockroach/pkg/sql/colexec/colbuilder.wrapRowSources({0x6755af0, 0xc00217b340}, 0x5154c3e, {0xc001d677a0, 0x1, 0x0}, {0xc003b38e88, 0x1, 0x7f33182d5a68}, 0xc003bcde00, ...)
	github.com/cockroachdb/cockroach/pkg/sql/colexec/colbuilder/execplan.go:109 +0x586 fp=0xc0025436e0 sp=0xc002543558 pc=0x3140ee6
github.com/cockroachdb/cockroach/pkg/sql/colexec/colbuilder.opResult.createAndWrapRowSource({0x100000000000000}, {0x6755af0, 0xc00217b340}, 0xc0029b4f00, 0xc0010b2380, {0xc001d677a0, 0x1, 0x1}, {0xc003b38e88, 0x1, ...}, ...)
	github.com/cockroachdb/cockroach/pkg/sql/colexec/colbuilder/execplan.go:539 +0x31b fp=0xc002543810 sp=0xc0025436e0 pc=0x31438db
github.com/cockroachdb/cockroach/pkg/sql/colexec/colbuilder.NewColOperator({0x6755af0, 0xc00217b340}, 0xc0029b4f00, 0xc0010b2380)
	github.com/cockroachdb/cockroach/pkg/sql/colexec/colbuilder/execplan.go:689 +0x58cc fp=0xc002544a58 sp=0xc002543810 pc=0x314982c
github.com/cockroachdb/cockroach/pkg/sql/colflow.(*vectorizedFlowCreator).setupFlow.func1()
	github.com/cockroachdb/cockroach/pkg/sql/colflow/vectorized_flow.go:1130 +0x86d fp=0xc002544e98 sp=0xc002544a58 pc=0x316422d
github.com/cockroachdb/cockroach/pkg/sql/colexecerror.CatchVectorizedRuntimeError(0x0)
	github.com/cockroachdb/cockroach/pkg/sql/colexecerror/error.go:91 +0x62 fp=0xc002544ed8 sp=0xc002544e98 pc=0x2017462
github.com/cockroachdb/cockroach/pkg/sql/colflow.(*vectorizedFlowCreator).setupFlow(0xc004b60700, {0x6755af0, 0xc00217b340}, 0xc000e80000, {0xc002b03100, 0xc003429680, 0x0}, {0xc001a3a940, 0x2, 0x2}, ...)
	github.com/cockroachdb/cockroach/pkg/sql/colflow/vectorized_flow.go:1069 +0xe5 fp=0xc002544f68 sp=0xc002544ed8 pc=0x31638c5
github.com/cockroachdb/cockroach/pkg/sql/colflow.(*vectorizedFlow).Setup(0xc003429680, {0x6755b98, 0xc003bcda10}, 0xc002444870, 0x80)
	github.com/cockroachdb/cockroach/pkg/sql/colflow/vectorized_flow.go:221 +0x511 fp=0xc0025451a0 sp=0xc002544f68 pc=0x315e091
github.com/cockroachdb/cockroach/pkg/sql/distsql.(*ServerImpl).setupFlow(0xc002025200, {0x6755b98, 0xc003ee01e0}, 0xc003bf6000, 0x0, 0xc002444840, {0x66cedb8, 0xc000d37880}, {0x66ced90, 0xc000d37880}, ...)
	github.com/cockroachdb/cockroach/pkg/sql/distsql/server.go:401 +0x1036 fp=0xc002545768 sp=0xc0025451a0 pc=0x31789d6
github.com/cockroachdb/cockroach/pkg/sql/distsql.(*ServerImpl).SetupLocalSyncFlow(0x200000000, {0x6755b98, 0xc003ee01e0}, 0x800000001, 0x4fed307, {0x66cedb8, 0xc000d37880}, {0x66ced90, 0xc000d37880}, {0xc004a98a80, ...})
	github.com/cockroachdb/cockroach/pkg/sql/distsql/server.go:563 +0x106 fp=0xc002545820 sp=0xc002545768 pc=0x3179ee6
github.com/cockroachdb/cockroach/pkg/sql.(*DistSQLPlanner).setupFlows(0xc001fb4b40, {0x6755b98, 0xc003ee01e0}, 0xc004a98a80, 0x0, 0xc003bcd9b0, 0xc000d37880, {0xc004a98a80, 0xc004ca02c8, 0x1, ...}, ...)
	github.com/cockroachdb/cockroach/pkg/sql/distsql_running.go:374 +0x8a6 fp=0xc002545cb0 sp=0xc002545820 pc=0x32acee6
github.com/cockroachdb/cockroach/pkg/sql.(*DistSQLPlanner).Run(0xc001fb4b40, {0x6755b98, 0xc003ee01e0}, 0xc0010b2000, 0xc004752460, 0xc0033fed80, 0xc000d37880, 0xc004a98a80, 0x0)
	github.com/cockroachdb/cockroach/pkg/sql/distsql_running.go:565 +0x7b4 fp=0xc0025463c0 sp=0xc002545cb0 pc=0x32ad754
github.com/cockroachdb/cockroach/pkg/sql.(*DistSQLPlanner).planAndRunPostquery(0xc001fb4b40, {0x6755b98, 0xc003ee01e0}, {{0x6757530, 0xc004715b80}, 0x0}, 0xc004ca0718, 0xc004a98a80, 0x5089670)
	github.com/cockroachdb/cockroach/pkg/sql/distsql_running.go:1662 +0x58c fp=0xc002546528 sp=0xc0025463c0 pc=0x32b2cac
github.com/cockroachdb/cockroach/pkg/sql.(*DistSQLPlanner).PlanAndRunCascadesAndChecks(0xc002b55180, {0x6755b98, 0xc003ee01e0}, 0xc004ca0718, 0xc0025467b8, 0xc004ca0da0, 0xc00169b040)
	github.com/cockroachdb/cockroach/pkg/sql/distsql_running.go:1593 +0xb95 fp=0xc002546690 sp=0xc002546528 pc=0x32b2595
github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).execWithDistSQLEngine(0xc004ca0000, {0x6755b98, 0xc003ee01e0}, 0xc004ca0718, 0xc003ee01e0, {0x7f32eddf1270, 0xc0040edbc0}, 0x37f, 0xc003514e98)
	github.com/cockroachdb/cockroach/pkg/sql/conn_executor_exec.go:1495 +0x78f fp=0xc002546808 sp=0xc002546690 pc=0x323b6cf
github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).dispatchToExecutionEngine(0xc004ca0000, {0x6755b98, 0xc00be14990}, 0xc004ca0718, {0x7f32eddf1270, 0xc0040edbc0})
	github.com/cockroachdb/cockroach/pkg/sql/conn_executor_exec.go:1159 +0xb87 fp=0xc002546c08 sp=0xc002546808 pc=0x32390e7
github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).execStmtInOpenState(0xc004ca0000, {0x6755af0, 0xc0018e9d80}, {{0x67866b8, 0xc000ec5f90}, {0xc0012469f0, 0x5c0}, 0x0, 0x2}, 0x0, ...)
	github.com/cockroachdb/cockroach/pkg/sql/conn_executor_exec.go:690 +0x2091 fp=0xc0025474e0 sp=0xc002546c08 pc=0x3234fb1
github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).execStmt(0xc004ca0000, {0x6755af0, 0xc0018e9d80}, {{0x67866b8, 0xc000ec5f90}, {0xc0012469f0, 0x5c0}, 0x0, 0x2}, 0x0, ...)
	github.com/cockroachdb/cockroach/pkg/sql/conn_executor_exec.go:145 +0x59e fp=0xc002547750 sp=0xc0025474e0 pc=0x323207e
github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).execCmd.func1({{{0x67866b8, 0xc000ec5f90}, {0xc0012469f0, 0x5c0}, 0x0, 0x2}, {0xc0ab60f913856e09, 0xc66696ab7, 0x0}, {0xc0ab60f91385731d, ...}, ...}, ...)
	github.com/cockroachdb/cockroach/pkg/sql/conn_executor.go:1911 +0x2f6 fp=0xc002547808 sp=0xc002547750 pc=0x3227916
github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).execCmd(0xc004ca0000)
	github.com/cockroachdb/cockroach/pkg/sql/conn_executor.go:1915 +0xb48 fp=0xc002547cd0 sp=0xc002547808 pc=0x3225ea8
github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).run(0xc004ca0000, {0x6755af0, 0xc004d46f00}, 0xc000497f80, 0x0, 0xc004c921c0)
	github.com/cockroachdb/cockroach/pkg/sql/conn_executor.go:1837 +0x216 fp=0xc002547d88 sp=0xc002547cd0 pc=0x3224fb6
github.com/cockroachdb/cockroach/pkg/sql.(*Server).ServeConn(0xc0004a0150, {0x6755af0, 0xc004d46f00}, {0xc00155ea00}, 0x3, 0xc003ed10f8)
	github.com/cockroachdb/cockroach/pkg/sql/conn_executor.go:778 +0xfb fp=0xc002547e38 sp=0xc002547d88 pc=0x3220c5b
github.com/cockroachdb/cockroach/pkg/sql/pgwire.(*conn).processCommandsAsync.func1()
	github.com/cockroachdb/cockroach/pkg/sql/pgwire/conn.go:727 +0x3fe fp=0xc002547fe0 sp=0xc002547e38 pc=0x360367e
runtime.goexit()
	GOROOT/src/runtime/asm_amd64.s:1581 +0x1 fp=0xc002547fe8 sp=0xc002547fe0 pc=0x4bf321
created by github.com/cockroachdb/cockroach/pkg/sql/pgwire.(*conn).processCommandsAsync
	github.com/cockroachdb/cockroach/pkg/sql/pgwire/conn.go:638 +0x232

```

--- a/pkg/sql/colfetcher/index_join.go
+++ b/pkg/sql/colfetcher/index_join.go
@@ -483,9 +483,10 @@ func NewColIndexJoin(
 	cFetcherMemoryLimit := totalMemoryLimit
 
 	var kvFetcher *row.KVFetcher
-	useStreamer := flowCtx.Txn != nil && flowCtx.Txn.Type() == kv.LeafTxn &&
-		flowCtx.MakeLeafTxn != nil && row.CanUseStreamer(ctx, flowCtx.EvalCtx.Settings)
-	txn := flowCtx.Txn
+	useStreamer, txn, err := flowCtx.UseStreamer()
+	if err != nil {
+		return nil, err
+	}
 	if useStreamer {
 		if streamerBudgetAcc == nil {
 			return nil, errors.AssertionFailedf("streamer budget account is nil when the Streamer API is desired")
@@ -497,10 +498,6 @@ func NewColIndexJoin(
 		// and we'll give the remaining memory to the streamer budget below.
 		cFetcherMemoryLimit = int64(math.Ceil(float64(totalMemoryLimit) / 8.0))
 		streamerBudgetLimit := 7 * cFetcherMemoryLimit
-		txn, err = flowCtx.MakeLeafTxn()
-		if err != nil {
-			return nil, err
-		}
 		kvFetcher = row.NewStreamingKVFetcher(
 			flowCtx.Cfg.DistSender,
 			flowCtx.Stopper(),

--- a/pkg/sql/distsql_running.go
+++ b/pkg/sql/distsql_running.go
@@ -37,7 +37,6 @@ import (
 	"github.com/cockroachdb/cockroach/pkg/sql/pgwire/pgcode"
 	"github.com/cockroachdb/cockroach/pkg/sql/pgwire/pgerror"
 	"github.com/cockroachdb/cockroach/pkg/sql/physicalplan"
-	"github.com/cockroachdb/cockroach/pkg/sql/row"
 	"github.com/cockroachdb/cockroach/pkg/sql/rowenc"
 	"github.com/cockroachdb/cockroach/pkg/sql/rowexec"
 	"github.com/cockroachdb/cockroach/pkg/sql/sem/tree"
@@ -477,7 +476,7 @@ func (dsp *DistSQLPlanner) Run(
 			// given a LeafTxn. In order for that LeafTxn to be created later,
 			// during the flow setup, we need to populate leafInputState below,
 			// so we tell the localState that there is concurrency.
-			if row.CanUseStreamer(ctx, dsp.st) {
+			if execinfra.CanUseStreamer(dsp.st) {
 				for _, proc := range plan.Processors {
 					if jr := proc.Spec.Core.JoinReader; jr != nil {
 						// Both index and lookup joins, with and without

--- a/pkg/sql/execinfra/BUILD.bazel
+++ b/pkg/sql/execinfra/BUILD.bazel
@@ -37,6 +37,7 @@ go_library(
         "//pkg/roachpb",
         "//pkg/rpc",
         "//pkg/rpc/nodedialer",
+        "//pkg/settings",
         "//pkg/settings/cluster",
         "//pkg/sql/catalog",
         "//pkg/sql/catalog/descpb",

--- a/pkg/sql/execinfra/readerbase.go
+++ b/pkg/sql/execinfra/readerbase.go
@@ -16,8 +16,11 @@ import (
 	"strings"
 
 	"github.com/cockroachdb/cockroach/pkg/keys"
+	"github.com/cockroachdb/cockroach/pkg/kv"
 	"github.com/cockroachdb/cockroach/pkg/kv/kvclient/rangecache"
 	"github.com/cockroachdb/cockroach/pkg/roachpb"
+	"github.com/cockroachdb/cockroach/pkg/settings"
+	"github.com/cockroachdb/cockroach/pkg/settings/cluster"
 	"github.com/cockroachdb/cockroach/pkg/sql/execinfrapb"
 	"github.com/cockroachdb/cockroach/pkg/util/log"
 	"github.com/cockroachdb/errors"
@@ -199,3 +202,37 @@ func (h *LimitHintHelper) ReadSomeRows(rowsRead int64) error {
 	}
 	return nil
 }
+
+// CanUseStreamer returns whether the kvstreamer.Streamer API should be used if
+// possible.
+func CanUseStreamer(settings *cluster.Settings) bool {
+	return useStreamerEnabled.Get(&settings.SV)
+}
+
+// UseStreamer returns whether the kvstreamer.Streamer API should be used as
+// well as the txn that should be used (regardless of the boolean return value).
+func (flowCtx *FlowCtx) UseStreamer() (bool, *kv.Txn, error) {
+	useStreamer := CanUseStreamer(flowCtx.EvalCtx.Settings) && flowCtx.Txn != nil &&
+		flowCtx.Txn.Type() == kv.LeafTxn && flowCtx.MakeLeafTxn != nil
+	if !useStreamer {
+		return false, flowCtx.Txn, nil
+	}
+	leafTxn, err := flowCtx.MakeLeafTxn()
+	if leafTxn == nil || err != nil {
+		// leafTxn might be nil in some flows which run outside of the txn, the
+		// streamer should not be used in such cases.
+		return false, flowCtx.Txn, err
+	}
+	return true, leafTxn, nil
+}
+
+// useStreamerEnabled determines whether the Streamer API should be used.
+// TODO(yuzefovich): remove this in 23.1.
+var useStreamerEnabled = settings.RegisterBoolSetting(
+	settings.TenantReadOnly,
+	"sql.distsql.use_streamer.enabled",
+	"determines whether the usage of the Streamer API is allowed. "+
+		"Enabling this will increase the speed of lookup/index joins "+
+		"while adhering to memory limits.",
+	true,
+)

--- a/pkg/sql/row/kv_batch_streamer.go
+++ b/pkg/sql/row/kv_batch_streamer.go
@@ -16,29 +16,11 @@ import (
 	"github.com/cockroachdb/cockroach/pkg/kv/kvclient/kvstreamer"
 	"github.com/cockroachdb/cockroach/pkg/kv/kvserver/concurrency/lock"
 	"github.com/cockroachdb/cockroach/pkg/roachpb"
-	"github.com/cockroachdb/cockroach/pkg/settings"
-	"github.com/cockroachdb/cockroach/pkg/settings/cluster"
 	"github.com/cockroachdb/cockroach/pkg/sql/catalog/descpb"
 	"github.com/cockroachdb/cockroach/pkg/sql/rowinfra"
 	"github.com/cockroachdb/cockroach/pkg/util/log"
 	"github.com/cockroachdb/cockroach/pkg/util/mon"
 	"github.com/cockroachdb/errors"
-)
-
-// CanUseStreamer returns whether the kvstreamer.Streamer API should be used.
-func CanUseStreamer(ctx context.Context, settings *cluster.Settings) bool {
-	return useStreamerEnabled.Get(&settings.SV)
-}
-
-// useStreamerEnabled determines whether the Streamer API should be used.
-// TODO(yuzefovich): remove this in 23.1.
-var useStreamerEnabled = settings.RegisterBoolSetting(
-	settings.TenantReadOnly,
-	"sql.distsql.use_streamer.enabled",
-	"determines whether the usage of the Streamer API is allowed. "+
-		"Enabling this will increase the speed of lookup/index joins "+
-		"while adhering to memory limits.",
-	true,
 )
 
 // txnKVStreamer handles retrieval of key/values.

--- a/pkg/sql/rowexec/joinreader.go
+++ b/pkg/sql/rowexec/joinreader.go
@@ -302,8 +302,10 @@ func newJoinReader(
 	if flowCtx.EvalCtx.SessionData().ParallelizeMultiKeyLookupJoinsEnabled {
 		shouldLimitBatches = false
 	}
-	useStreamer := flowCtx.Txn != nil && flowCtx.Txn.Type() == kv.LeafTxn &&
-		flowCtx.MakeLeafTxn != nil && row.CanUseStreamer(flowCtx.EvalCtx.Ctx(), flowCtx.EvalCtx.Settings)
+	useStreamer, txn, err := flowCtx.UseStreamer()
+	if err != nil {
+		return nil, err
+	}
 
 	jr := &joinReader{
 		fetchSpec:                         spec.FetchSpec,
@@ -314,22 +316,13 @@ func newJoinReader(
 		outputGroupContinuationForLeftRow: spec.OutputGroupContinuationForLeftRow,
 		shouldLimitBatches:                shouldLimitBatches,
 		readerType:                        readerType,
+		txn:                               txn,
 		usesStreamer:                      useStreamer,
 		lookupBatchBytesLimit:             rowinfra.BytesLimit(spec.LookupBatchBytesLimit),
 		limitHintHelper:                   execinfra.MakeLimitHintHelper(spec.LimitHint, post),
 	}
 	if readerType != indexJoinReaderType {
 		jr.groupingState = &inputBatchGroupingState{doGrouping: spec.LeftJoinWithPairedJoiner}
-	}
-
-	if useStreamer {
-		var err error
-		jr.txn, err = flowCtx.MakeLeafTxn()
-		if err != nil {
-			return nil, err
-		}
-	} else {
-		jr.txn = flowCtx.Txn
 	}
 
 	// Make sure the key column types are hydrated. The fetched column types will

